evm/src/logic.rs

```diff
@@ -78,62 +79,43 @@ fn check_op_flags<F: RichField>(lv: &[F; columns::NUM_COLUMNS]) -> Op {
     assert!(is_or <= 1);
     let is_xor = lv[columns::IS_XOR].to_canonical_u64();
     assert!(is_xor <= 1);
-    assert_eq!(is_and + is_or + is_xor, 1);
+    assert!(is_and + is_or + is_xor <= 1);
     if is_and == 1 {
         Op::And
     } else if is_or == 1 {
         Op::Or
     } else if is_xor == 1 {
         Op::Xor
     } else {
-        panic!("unknown operation")
+        Op::Zero
     }
 }
 
-fn check_limb_length<F: RichField>(lv: &[F; columns::NUM_COLUMNS]) {
-    for (packed_input_cols, bit_cols) in [
-        (columns::INPUT0_PACKED, columns::INPUT0_BITS),
-        (columns::INPUT1_PACKED, columns::INPUT1_BITS),
-    ] {
-        let limb_bit_cols_iter = columns::limb_bit_cols_for_input(bit_cols);
-        // Not actually reading/writing the bit columns, but this is a convenient way of
-        // calculating the size of each limb.
-        for (packed_limb_col, limb_bit_cols) in packed_input_cols.zip(limb_bit_cols_iter) {
-            let packed_limb = lv[packed_limb_col].to_canonical_u64();
-            let limb_length = limb_bit_cols.end - limb_bit_cols.start;
-            assert_eq!(packed_limb >> limb_length, 0);
-        }
-    }
-}
-
-fn make_bit_decomposition<F: RichField>(lv: &mut [F; columns::NUM_COLUMNS]) {
-    for (packed_input_cols, bit_cols) in [
-        (columns::INPUT0_PACKED, columns::INPUT0_BITS),
-        (columns::INPUT1_PACKED, columns::INPUT1_BITS),
-    ] {
-        for (i, limb_col) in packed_input_cols.enumerate() {
-            let limb = lv[limb_col].to_canonical_u64();
-            let limb_bits_cols = bit_cols
-                .clone()
-                .skip(i * PACKED_LIMB_BITS)
-                .take(PACKED_LIMB_BITS);
-            for (j, col) in limb_bits_cols.enumerate() {
-                let bit = (limb >> j) & 1;
-                lv[col] = F::from_canonical_u64(bit);
-            }
+fn check_bits<F: RichField>(lv: &[F; columns::NUM_COLUMNS]) {
+    for bit_cols in [columns::INPUT0, columns::INPUT1] {
+        for bit_col in bit_cols {
+            let bit = lv[bit_col].to_canonical_u64();
+            assert!(bit <= 1);
         }
     }
 }
 
 fn make_result<F: RichField>(lv: &mut [F; columns::NUM_COLUMNS], op: Op) {
-    for (res_col, limb_in0_col, limb_in1_col) in izip!(
+    for (res_col, limb_in0_cols, limb_in1_cols) in izip!(
         columns::RESULT,
-        columns::INPUT0_PACKED,
-        columns::INPUT1_PACKED
+        columns::limb_bit_cols_for_input(columns::INPUT0),
+        columns::limb_bit_cols_for_input(columns::INPUT1),
     ) {
-        let limb_in0 = lv[limb_in0_col].to_canonical_u64();
-        let limb_in1 = lv[limb_in1_col].to_canonical_u64();
+        let limb_in0: u64 = limb_in0_cols
+            .enumerate()
+            .map(|(i, col)| lv[col].to_canonical_u64() << i)
+            .sum();
```

Comment:
nit: It seems like this conversion `bits -> u64/usize` is done in a few places. Maybe we could add a function for it.